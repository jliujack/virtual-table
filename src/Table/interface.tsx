import * as React from 'react';

import { TooltipProps, CheckboxProps, PaginationProps } from 'antd';

import {
  GetRowKey,
  ColumnType as RcColumnType,
  RenderedCell as RcRenderedCell,
  ExpandableConfig,
  CellRenderProps as RcCellRenderProps,
  DefaultRecordType,
  TableFunc,
} from './inner-table/interface';
import { INTERNAL_SELECTION_ITEM } from './hooks/useSelection';
import { tuple } from '../_util/type';
// import { TableAction } from './Table';

export { GetRowKey, ExpandableConfig, DefaultRecordType, TableFunc };

export type Key = React.Key;

export type RowSelectionType = 'checkbox' | 'radio';

export type SelectionItemSelectFn = (currentRowKeys: Key[]) => void;

export type ExpandType = null | 'row' | 'nest';

export interface TableLocale {
  filterTitle?: string;
  filterConfirm?: React.ReactNode;
  filterReset?: React.ReactNode;
  filterEmptyText?: React.ReactNode;
  emptyText?: React.ReactNode | (() => React.ReactNode);
  selectAll?: React.ReactNode;
  selectNone?: React.ReactNode;
  selectInvert?: React.ReactNode;
  selectionAll?: React.ReactNode;
  sortTitle?: string;
  expand?: string;
  collapse?: string;
  triggerDesc?: string;
  triggerAsc?: string;
  cancelSort?: string;
}

export type SortOrder = 'desc' | 'asc' | null;

const TableActions = tuple('paginate', 'sort', 'filter');
export type TableAction = typeof TableActions[number];

export type CompareFn<T> = (a: T, b: T, sortOrder?: SortOrder) => number;

export interface ColumnFilterItem {
  text: React.ReactNode;
  value: string | number | boolean;
  children?: ColumnFilterItem[];
}

export interface ColumnTitleProps<RecordType> {
  /** @deprecated Please use `sorterColumns` instead. */
  sortOrder?: SortOrder;
  /** @deprecated Please use `sorterColumns` instead. */
  sortColumn?: ColumnType<RecordType>;
  sortColumns?: { column: ColumnType<RecordType>; order: SortOrder }[];

  filters?: Record<string, string[]>;
}

export type ColumnTitle<RecordType> =
  | React.ReactNode
  | ((props: ColumnTitleProps<RecordType>) => React.ReactNode);

export type FilterValue = (Key | boolean)[];
export type FilterKey = Key[] | null;
export interface FilterConfirmProps {
  closeDropdown: boolean;
}

export interface FilterDropdownProps {
  prefixCls: string;
  setSelectedKeys: (selectedKeys: React.Key[]) => void;
  selectedKeys: React.Key[];
  confirm: (param?: FilterConfirmProps) => void;
  clearFilters?: () => void;
  filters?: ColumnFilterItem[];
  visible: boolean;
}

export interface ColumnType<RecordType> extends RcColumnType<RecordType> {
  /**
   * @description       表头内容，可以是React节点
   */
  title?: ColumnTitle<RecordType>;
  // Sorter
  /**
   * @description       设置当前表头字段是否排序或者排序的计算函数，如是多行排序需要设置为对象并添加排序优先级
   */
  sorter?:
    | boolean
    | CompareFn<RecordType>
    | {
        compare?: CompareFn<RecordType>;
        /** Config multiple sorter order priority */
        multiple?: number;
      };
  /**
   * @description       表明当前表头的排序方向
   */
  sortOrder?: SortOrder;
  defaultSortOrder?: SortOrder;
  sortDirections?: SortOrder[];
  showSorterTooltip?: boolean | TooltipProps;

  // Filter
  filtered?: boolean;
  filters?: ColumnFilterItem[];
  filterDropdown?: React.ReactNode | ((props: FilterDropdownProps) => React.ReactNode);
  filterMultiple?: boolean;
  filteredValue?: FilterValue | null;
  defaultFilteredValue?: FilterValue | null;
  filterIcon?: React.ReactNode | ((filtered: boolean) => React.ReactNode);
  onFilter?: (value: string | number | boolean, record: RecordType) => boolean;
  filterDropdownVisible?: boolean;
  onFilterDropdownVisibleChange?: (visible: boolean) => void;
}

export interface ColumnGroupType<RecordType> extends Omit<ColumnType<RecordType>, 'dataIndex'> {
  children: ColumnsType<RecordType>;
}

export type ColumnsType<RecordType = unknown> = (
  | ColumnGroupType<RecordType>
  | ColumnType<RecordType>
)[];

// ====================== Cell ======================
export type CellRenderProps<RecordType> = RcCellRenderProps<RecordType>;
export interface SelectionItem {
  key: string;
  text: React.ReactNode;
  onSelect?: SelectionItemSelectFn;
}

export type SelectionSelectFn<T> = (
  record: T,
  selected: boolean,
  selectedRows: T[],
  nativeEvent: Event,
) => void;

export interface TableRowSelection<T> {
  /** Keep the selection keys in list even the key not exist in `dataSource` anymore */
  preserveSelectedRowKeys?: boolean;
  type?: RowSelectionType;
  selectedRowKeys?: Key[];
  defaultSelectedRowKeys?: Key[];
  onChange?: (selectedRowKeys: Key[], selectedRows: T[]) => void;
  getCheckboxProps?: ({
    rowData,
  }: {
    rowData: T;
  }) => Partial<Omit<CheckboxProps, 'checked' | 'defaultChecked'>>;
  onSelect?: SelectionSelectFn<T>;
  onSelectMultiple?: (selected: boolean, selectedRows: T[], changeRows: T[]) => void;
  /** @deprecated This function is meaningless and should use `onChange` instead */
  onSelectAll?: (selected: boolean, selectedRows: T[], changeRows: T[]) => void;
  /** @deprecated This function is meaningless and should use `onChange` instead */
  onSelectInvert?: (selectedRowKeys: Key[]) => void;
  onSelectNone?: () => void;
  selections?: INTERNAL_SELECTION_ITEM[] | boolean;
  hideSelectAll?: boolean;
  fixed?: boolean;
  columnWidth?: number;
  columnTitle?: string | React.ReactNode;
  checkStrictly?: boolean;
  renderCell?: (
    value: boolean,
    record: T,
    index: number,
    originNode: React.ReactNode,
  ) => React.ReactNode | RcRenderedCell<T>;
}

export type TransformColumns<RecordType> = (
  columns: ColumnsType<RecordType>,
) => ColumnsType<RecordType>;

export interface TableCurrentDataSource<RecordType> {
  currentDataSource: RecordType[];
  action: TableAction;
}

export interface SorterResult<RecordType> {
  column: ColumnType<RecordType>;
  order: SortOrder;
  field?: Key | readonly Key[];
  /** @description 生成的key, 优先级为column.key??column.dataIndex??根据column顺序生成 */
  key: Key;
  /** @description 列的key */
  columnKey?: Key;
}

export type GetPopupContainer = (triggerNode: HTMLElement) => HTMLElement;

type TablePaginationPosition =
  | 'topLeft'
  | 'topCenter'
  | 'topRight'
  | 'bottomLeft'
  | 'bottomCenter'
  | 'bottomRight';

export interface TablePaginationConfig extends PaginationProps {
  position?: TablePaginationPosition[];
}

export interface ResizableConfig {
  prefixCls: string;
  onColumnResize?: (widthChange: Record<Key, number>) => void;
  minColumnWidth: number;
}
