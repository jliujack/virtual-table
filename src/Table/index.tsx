import Table, { TableProps, TableRef } from './Table';
import { ColumnProps } from './Column';
import type {
  ColumnsType,
  ColumnType,
  ColumnGroupType,
  ExpandableConfig,
  TableRowSelection,
} from './interface';

export type {
  TableProps,
  TableRef,
  ColumnProps,
  ColumnsType,
  ColumnType,
  ColumnGroupType,
  ExpandableConfig,
  TableRowSelection,
};

export default Table;

/** Description of the TableProps */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function TablePropsApi<RecordType>(props: TableProps<RecordType>) {
  console.log('liujie log:', 'props:', props);
  return null;
}

/** Description of the Columns */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function ColumnsApi<RecordType>(props: ColumnType<RecordType>) {
  console.log('liujie log:', 'props:', props);
  return null;
}

/** Description of the Expandable */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function Expandable<RecordType>(props: ExpandableConfig<RecordType>) {
  console.log('liujie log:', 'props:', props);
  return null;
}

/** Description of the Selection */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function Selection<RecordType>(props: TableRowSelection<RecordType>) {
  console.log('liujie log:', 'props:', props);
  return null;
}
