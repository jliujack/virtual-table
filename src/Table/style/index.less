@table-prefix-cls: QT-table;
@table-dark-prefix-cls: QT-dark-table;

@import '../../style/themes/index';
@import '../../style/mixins/index';

@import './color.less';

@import './bordered';

@import './table.less';
@import './table-dark.less';
@import './row-flag.less';
// @import './react-texty.less';
