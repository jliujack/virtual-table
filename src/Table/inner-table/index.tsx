import Table, { TableProps, DEFAULT_PROPS } from './Table';
import type { ColumnsType } from './interface';

import { getPathValue } from './utils/valueUtil';

export { getPathValue, DEFAULT_PROPS };
export type { ColumnsType, TableProps };
export default Table;
