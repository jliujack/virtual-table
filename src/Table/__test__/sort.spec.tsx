/* eslint-disable max-classes-per-file */
import React from 'react';
import { render, mount } from 'enzyme';
import Table, { TableProps, ColumnType } from '../index';

describe('Table.sort', () => {
  const sorterFn = (a, b) => a.name[0].charCodeAt() - b.name[0].charCodeAt();

  const column = {
    title: 'Name',
    dataIndex: 'name',
    key: 'name',
    sorter: sorterFn,
  };

  const data = [
    { key: 0, name: 'Jack' },
    { key: 1, name: 'Lucy' },
    { key: 2, name: 'Tom' },
    { key: 3, name: 'Jerry' },
  ];

  type Record = typeof data[number];

  function createTable(tableProps: TableProps<Record> = {}, columnProps: ColumnType<Record> = {}) {
    return (
      <Table
        columns={[
          {
            ...column,
            ...columnProps,
          },
        ]}
        dataSource={data}
        specialClass="bordered-spec"
        {...tableProps}
      />
    );
  }

  function renderedNames(wrapper) {
    return wrapper.find('BodyRow').map((row) => row.props().record.name);
  }

  it('renders sort icon correctly', () => {
    const wrapper = render(createTable());
    expect(wrapper.find('.QT-table-thead')).toMatchSnapshot();
  });

  it('default sort order asc', () => {
    const wrapper = mount(
      createTable(
        {},
        {
          defaultSortOrder: 'asc',
        },
      ),
    );

    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
  });
  it('default sort order desc', () => {
    const wrapper = mount(
      createTable(
        {},
        {
          defaultSortOrder: 'desc',
        },
      ),
    );

    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
  });

  it('sort records', () => {
    const wrapper = mount(createTable());

    // asc
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);

    // desc
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
  });

  describe('can be controlled by sortOrder', () => {
    it('single', () => {
      const wrapper = mount(
        createTable({
          columns: [{ ...column, sortOrder: 'asc' }],
        }),
      );
      expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
    });

    it('invalidate mix with single & multiple sorters', () => {
      const wrapper = mount(
        createTable({
          columns: [
            {
              title: 'Name',
              dataIndex: 'name',
              sortOrder: 'asc',
              sorter: {
                multiple: 1,
              },
            },
            {
              title: 'Name',
              dataIndex: 'name',
              sortOrder: 'asc',
              sorter: {},
            },
          ],
        }),
      );

      expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    });
  });

  it('provides sortOrder in the sorterFn', () => {
    let actualSortOrder;
    mount(
      createTable(
        {},
        {
          sortOrder: 'asc',
          sorter: (a, b, sortOrder) => {
            actualSortOrder = sortOrder;
            return sorterFn(a, b);
          },
        },
      ),
    );
    expect(actualSortOrder).toEqual('asc');
  });

  it('can update column sortOrder', () => {
    const wrapper = mount(
      createTable({
        columns: [column],
      }),
    );
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
    wrapper.setProps({
      columns: [{ ...column, sortOrder: 'asc' }],
    });
    wrapper.update();
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);
  });

  it('fires change event', () => {
    const handleChange = jest.fn();
    const wrapper = mount(createTable({ onChange: handleChange }));

    // ascent
    wrapper.find('.QT-table-column-sorters').simulate('click');
    const sorter1 = handleChange.mock.calls[0][2];
    expect(sorter1.column.dataIndex).toBe('name');
    expect(sorter1.order).toBe('asc');
    expect(sorter1.field).toBe('name');
    expect(sorter1.columnKey).toBe('name');

    wrapper.find('.QT-table-column-sorters').simulate('click');
    const sorter2 = handleChange.mock.calls[1][2];
    expect(sorter2.column.dataIndex).toBe('name');
    expect(sorter2.order).toBe('desc');
    expect(sorter2.field).toBe('name');
    expect(sorter2.columnKey).toBe('name');

    wrapper.find('.QT-table-column-sorters').simulate('click');
    const sorter3 = handleChange.mock.calls[2][2];
    expect(sorter3.column).toEqual({});
    expect(sorter3.order).toBe(undefined);
    expect(sorter3.field).toBe('name');
    expect(sorter3.columnKey).toBe('name');
  });

  // it('hover header show sorter tooltip', () => {
  //   // tooltip has delay
  //   jest.useFakeTimers();
  //   const wrapper = mount(createTable({}));
  //   // default show sorter tooltip
  //   wrapper.find('.QT-table-column-sorters').simulate('mouseenter');
  //   jest.runAllTimers();
  //   wrapper.update();
  //   expect(wrapper.find('.ant-tooltip-open').length).toBeTruthy();
  //   wrapper.find('.QT-table-column-sorters').simulate('mouseout');

  //   // set table props showSorterTooltip is false
  //   wrapper.setProps({ showSorterTooltip: false });
  //   jest.runAllTimers();
  //   wrapper.update();
  //   expect(wrapper.find('.ant-tooltip-open')).toHaveLength(0);
  //   // set table props showSorterTooltip is false, column showSorterTooltip is true
  //   wrapper.setProps({
  //     showSorterTooltip: false,
  //     columns: [{ ...column, showSorterTooltip: true }],
  //   });
  //   wrapper.find('.QT-table-column-sorters').simulate('mouseenter');
  //   jest.runAllTimers();
  //   wrapper.update();
  //   expect(wrapper.find('.ant-tooltip-open').length).toBeTruthy();
  //   wrapper.find('.QT-table-column-sorters').simulate('mouseout');
  //   // set table props showSorterTooltip is true, column showSorterTooltip is false
  //   wrapper.setProps({
  //     showSorterTooltip: true,
  //     columns: [{ ...column, showSorterTooltip: false }],
  //   });
  //   jest.runAllTimers();
  //   wrapper.update();
  //   expect(wrapper.find('.ant-tooltip-open')).toHaveLength(0);
  // });

  it('should show correct tooltip when showSorterTooltip is an object', () => {
    // basically copied from 'hover header show sorter tooltip'
    jest.useFakeTimers();
    const wrapper = mount(
      createTable({ showSorterTooltip: { placement: 'bottom', title: 'static title' } }),
    );
    wrapper.find('.QT-table-column-sorters').simulate('mouseenter');
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.ant-tooltip-open').length).toBeTruthy();
    wrapper.find('.QT-table-column-sorters').simulate('mouseout');

    wrapper.setProps({ showSorterTooltip: false });
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.ant-tooltip-open')).toHaveLength(0);
    wrapper.setProps({
      showSorterTooltip: false,
      columns: [{ ...column, showSorterTooltip: true }],
    });
    wrapper.find('.QT-table-column-sorters').simulate('mouseenter');
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.ant-tooltip-open').length).toBeTruthy();
    wrapper.find('.QT-table-column-sorters').simulate('mouseout');
    wrapper.setProps({
      showSorterTooltip: true,
      columns: [{ ...column, showSorterTooltip: false }],
    });
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.ant-tooltip-open')).toHaveLength(0);
  });

  it('works with grouping columns in controlled mode', () => {
    const columns = [
      {
        title: 'group',
        key: 'group',
        children: [
          {
            title: 'Name',
            dataIndex: 'name',
            key: 'name',
            sorter: sorterFn,
            sortOrder: 'desc',
          },
          {
            title: 'Age',
            dataIndex: 'age',
            key: 'age',
          },
        ],
      },
    ];
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    const wrapper = mount(<Table columns={columns} dataSource={testData} />);

    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);
  });

  // https://github.com/ant-design/ant-design/issues/11246#issuecomment-405009167
  it('Allow column title as render props with sortOrder argument', () => {
    const title = ({ sortOrder }) => <div className="custom-title">{sortOrder}</div>;
    const columns = [
      {
        title,
        key: 'group',
        sorter: true,
      },
    ];
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    const wrapper = mount(<Table columns={columns} dataSource={testData} />);
    expect(wrapper.find('.custom-title').text()).toEqual('');
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(wrapper.find('.custom-title').text()).toEqual('asc');
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(wrapper.find('.custom-title').text()).toEqual('desc');
  });

  // https://github.com/ant-design/ant-design/pull/12264#discussion_r218053034
  it('should sort from beginning state when toggle from different columns', () => {
    const columns = [
      {
        title: 'name',
        dataIndex: 'name',
        sorter: true,
      },
      {
        title: 'age',
        dataIndex: 'age',
        sorter: true,
      },
    ];
    const testData = [
      { key: 0, name: 'Jack', age: 11 },
      { key: 1, name: 'Lucy', age: 20 },
      { key: 2, name: 'Tom', age: 21 },
      { key: 3, name: 'Jerry', age: 22 },
    ];
    const wrapper = mount(<Table columns={columns} dataSource={testData} />);

    const getNameColumn = () => wrapper.find('.QT-table-column-has-sorters').at(0);
    const getAgeColumn = () => wrapper.find('.QT-table-column-has-sorters').at(1);
    const getNameIcon = (name) => getNameColumn().find(`.QT-table-column-sorter-${name}`).first();
    const getAgeIcon = (name) => getAgeColumn().find(`.QT-table-column-sorter-${name}`).first();

    // sort name
    getNameColumn().simulate('click');
    expect(getNameIcon('up').hasClass('active')).toBeTruthy();
    expect(getAgeIcon('up').hasClass('active')).toBeFalsy();

    // sort age
    getAgeColumn().simulate('click');
    expect(getNameIcon('up').hasClass('active')).toBeFalsy();
    expect(getAgeIcon('up').hasClass('active')).toBeTruthy();
  });

  it('should first sort by desc, then asc, then cancel sort', () => {
    const wrapper = mount(
      createTable({
        sortDirections: ['desc', 'asc'],
      }),
    );

    // desc
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);

    // asc
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Jerry', 'Lucy', 'Tom']);

    // cancel sort
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
  });

  it('should first sort by desc, then cancel sort', () => {
    const wrapper = mount(
      createTable({
        sortDirections: ['desc'],
      }),
    );

    // desc
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);

    // cancel sort
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
  });

  it('should first sort by desc, then cancel sort. (column prop)', () => {
    const wrapper = mount(
      createTable(
        {},
        {
          sortDirections: ['desc'],
        },
      ),
    );

    // desc
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Lucy', 'Jack', 'Jerry']);

    // cancel sort
    wrapper.find('.QT-table-column-sorters').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy', 'Tom', 'Jerry']);
  });

  it('should support onHeaderCell in sort column', () => {
    const onClick = jest.fn();
    const wrapper = mount(
      <Table columns={[{ title: 'title', onHeaderCell: () => ({ onClick }), sorter: true }]} />,
    );
    wrapper.find('.QT-table-column-title').simulate('click');
    expect(onClick).toHaveBeenCalled();
  });

  it('could sort data with children', () => {
    const wrapper = mount(
      createTable(
        {
          expandable: {
            defaultExpandAllRows: true,
          },
          dataSource: [
            {
              key: '1',
              name: 'Brown',
              children: [
                {
                  key: '2',
                  name: 'Zoe',
                },
                {
                  key: '3',
                  name: 'Mike',
                  children: [
                    {
                      key: '3-1',
                      name: 'Petter',
                    },
                    {
                      key: '3-2',
                      name: 'Alex',
                    },
                  ],
                },
                {
                  key: '4',
                  name: 'Green',
                },
              ],
            },
          ],
        },
        {
          defaultSortOrder: 'asc',
        },
      ),
    );

    expect(renderedNames(wrapper)).toEqual(['Brown', 'Green', 'Mike', 'Alex', 'Petter', 'Zoe']);
  });

  // https://github.com/ant-design/ant-design/issues/19443
  it('should not being inifinite loop when using Table.Column with sortOrder', () => {
    class Demo extends React.Component {
      componentDidMount() {
        this.setState({});
      }

      render() {
        return (
          <Table dataSource={[]}>
            <Table.Column title="Age" dataIndex="age" sorter sortOrder="asc" key="age" />
          </Table>
        );
      }
    }
    expect(() => {
      mount(<Demo />);
    }).not.toThrow();
  });

  it('should support defaultOrder in Column', () => {
    const wrapper = mount(
      <Table dataSource={[{ key: '1', age: 1 }]} specialClass="table-sort-default-order">
        <Table.Column title="Age" dataIndex="age" sorter defaultSortOrder="asc" key="age" />
      </Table>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  // https://github.com/ant-design/ant-design/issues/20096
  it('invalidate sorter should not display sorter button', () => {
    const wrapper = mount(
      <Table
        columns={[
          {
            title: 'Name',
            dataIndex: 'name',
            key: 'name',
            sorter: false,
          },
          {
            title: 'Age',
            dataIndex: 'age',
            key: 'age',
            sorter: null,
          },
          {
            title: 'Address',
            dataIndex: 'address',
            key: 'address',
            sorter: undefined,
          },
        ]}
      />,
    );

    expect(wrapper.find('.QT-table-column-sorter-inner')).toHaveLength(0);
  });

  // https://github.com/ant-design/ant-design/issues/21193
  it('table with sugar column', () => {
    const wrapper = mount(
      <Table>
        <Table.Column
          title="Chinese Score"
          dataIndex="chinese"
          sorter={{
            compare: (a, b) => a.chinese - b.chinese,
            multiple: 3,
          }}
        />
        <Table.Column
          title="Math Score"
          dataIndex="math"
          sorter={{
            compare: (a, b) => a.math - b.math,
            multiple: 2,
          }}
        />
      </Table>,
    );

    wrapper.find('.QT-table-column-title').first().simulate('click');

    expect(wrapper.find('.QT-table-column-sort')).toBeTruthy();
  });

  it('surger should support sorterOrder', () => {
    const wrapper = mount(
      <Table>
        <Table.Column key="name" title="Name" dataIndex="name" sortOrder="asc" sorter />
      </Table>,
    );

    expect(wrapper.find('.QT-table-column-sorter-up').last().hasClass('active')).toBeTruthy();
    expect(wrapper.find('.QT-table-column-sorter-down').last().hasClass('active')).toBeFalsy();
  });

  it('controlled multiple group', () => {
    const groupColumns = [
      {
        title: 'Math Score',
        dataIndex: 'math1',
        sortOrder: 'asc',
        sorter: { multiple: 1 },
        children: [
          {
            title: 'math',
            dataIndex: 'math',
          },
        ],
      },
      {
        title: 'English Score',
        dataIndex: 'english',
        sortOrder: 'desc',
        sorter: { multiple: 2 },
      },
    ];

    const groupData = [
      {
        key: '1',
        name: 'John Brown',
        chinese: 98,
        math: 60,
        english: 70,
      },
    ];

    const wrapper = mount(<Table columns={groupColumns} data={groupData} />);
    wrapper.update();
    expect(
      wrapper
        .find('.QT-table-column-sorter-full')
        .first()
        .find('.QT-table-column-sorter-up')
        .first()
        .hasClass('active'),
    ).toBeTruthy();
    expect(
      wrapper
        .find('.QT-table-column-sorter-full')
        .last()
        .find('.QT-table-column-sorter-down')
        .first()
        .hasClass('active'),
    ).toBeTruthy();
  });

  // it('onChange with correct sorter for multiple', () => {
  //   const groupColumns = [
  //     {
  //       title: 'Math Score',
  //       dataIndex: 'math',
  //       sorter: { multiple: 1 },
  //     },
  //     {
  //       title: 'English Score',
  //       dataIndex: 'english',
  //       sorter: { multiple: 2 },
  //     },
  //   ];

  //   const groupData = [
  //     {
  //       key: '1',
  //       name: 'John Brown',
  //       chinese: 98,
  //       math: 60,
  //       english: 70,
  //     },
  //   ];

  //   const onChange = jest.fn();
  //   const wrapper = mount(<Table columns={groupColumns} data={groupData} onChange={onChange} />);

  //   function clickToMatchExpect(index, sorter) {
  //     wrapper.find('.QT-table-column-sorters').at(index).simulate('click');

  //     expect(onChange).toHaveBeenCalledWith(
  //       expect.anything(),
  //       expect.anything(),
  //       expect.objectContaining(sorter),
  //       expect.anything(),
  //     );

  //     onChange.mockReset();
  //   }

  //   // First
  //   clickToMatchExpect(0, { field: 'math', order: 'asc' });
  //   clickToMatchExpect(0, { field: 'math', order: 'desc' });
  //   clickToMatchExpect(0, { field: 'math', order: undefined });

  //   // Last
  //   clickToMatchExpect(1, { field: 'english', order: 'asc' });
  //   clickToMatchExpect(1, { field: 'english', order: 'desc' });
  //   clickToMatchExpect(1, { field: 'english', order: undefined });
  // });
});
