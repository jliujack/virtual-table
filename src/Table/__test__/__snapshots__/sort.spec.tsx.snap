// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.sort renders sort icon correctly 1`] = `
<div
  class="QT-table-thead"
  style="width:400px;height:36px"
>
  <div
    class="QT-table-header-row"
    style="height:36px"
  >
    <div
      class="QT-table-cell QT-table-column-has-sorters"
      style="position:absolute;left:0;width:400px;height:36px"
      title=""
    >
      <div
        class="QT-table-column-sorters"
      >
        <span
          class="QT-table-column-title"
        >
          Name
        </span>
        <span
          class="QT-table-column-sorter QT-table-column-sorter-full"
        >
          <span
            class="QT-table-column-sorter-inner"
          >
            <span
              aria-label="caret-up"
              class="anticon anticon-caret-up QT-table-column-sorter-up"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="caret-up"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                />
              </svg>
            </span>
            <span
              aria-label="caret-down"
              class="anticon anticon-caret-down QT-table-column-sorter-down"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="caret-down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.sort should support defaultOrder in Column 1`] = `
<div
  class="QT-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="QT-table QT-table-bordered"
      >
        <div
          class="QT-table-container"
        >
          <div
            class="QT-table-header"
            style="overflow: hidden; width: 400px;"
          >
            <div>
              <div
                class="QT-table-thead"
                style="width: 400px; height: 36px;"
              >
                <div
                  class="QT-table-header-row"
                  style="height: 36px;"
                >
                  <div
                    class="QT-table-cell QT-table-column-sort QT-table-column-has-sorters"
                    style="position: absolute; left: 0px; width: 400px; height: 36px;"
                    title=""
                  >
                    <div
                      class="QT-table-column-sorters"
                    >
                      <span
                        class="QT-table-column-title"
                      >
                        Age
                      </span>
                      <span
                        class="QT-table-column-sorter QT-table-column-sorter-full"
                      >
                        <span
                          class="QT-table-column-sorter-inner"
                        >
                          <span
                            aria-label="caret-up"
                            class="anticon anticon-caret-up QT-table-column-sorter-up active"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="caret-up"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="0 0 1024 1024"
                              width="1em"
                            >
                              <path
                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                              />
                            </svg>
                          </span>
                          <span
                            aria-label="caret-down"
                            class="anticon anticon-caret-down QT-table-column-sorter-down"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="caret-down"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="0 0 1024 1024"
                              width="1em"
                            >
                              <path
                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="table_special_aj7jq QT-table-tbody"
            style="position: relative; height: 364px; width: 400px; overflow: auto; will-change: unset; direction: ltr; overflow-x: hidden; overflow-y: hidden;"
          >
            <div
              style="height: 32px; width: 400px;"
            >
              <div
                class="QT-table-row-wrapper"
                style="position: absolute; left: 0px; top: 0px; height: 32px; width: 400px;"
              >
                <div
                  class="QT-table-row QT-table-row-level-0"
                  data-row-key="1"
                  style="height: 32px;"
                >
                  <div
                    class="QT-table-cell QT-table-column-sort QT-table__row-cell--with-flag"
                    style="position: absolute; left: 0px; width: 400px; height: 32px;"
                    title=""
                  >
                    1
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
